README.md
# lazymention

[![Build Status](https://travis-ci.org/strugee/lazymention.svg?branch=master)](https://travis-ci.org/strugee/lazymention)
[![Coverage Status](https://coveralls.io/repos/github/strugee/lazymention/badge.svg?branch=master)](https://coveralls.io/github/strugee/lazymention?branch=master)
[![NSP Status](https://nodesecurity.io/orgs/strugee/projects/530261ab-11a8-4f24-8b9f-55c005d0424e/badge)](https://nodesecurity.io/orgs/strugee/projects/530261ab-11a8-4f24-8b9f-55c005d0424e)
[![Greenkeeper badge](https://badges.greenkeeper.io/strugee/lazymention.svg)](https://greenkeeper.io/)

tl;dr: Support WebSub and Webmention on a static site

lazymention is a daemon that you run on a server. When you publish to your static [IndieWeb][] site, you ping lazymention, and it will poll your site, parse the microformats2 markup, and send Webmentions and WebSub notifications for content it hasn't seen before.

Webmention functionality is currently done. WebSub is planned for the future.

## Installation

    $ npm install -g --production lazymention

## Limitations

lazymention makes some maintainability tradeoffs that may result in denial-of-service problems if you allow input from sites you don't trust. Specifically:

* lazymention buffers all HTML pages in memory before parsing microformats2. In the future it may also retrieve multiple HTML pages per job. This may lead to high memory usage.
* lazymention's handling of microformats2 structures within pages is rather brittle. For example, if you mark an element as both an `h-entry` _and_ an `h-feed`, lazymention will probably crash. I will take PRs to fix these problems, especially those caused by nonmalicious markup, but in the meantime your process will be very dead.
* lazymention does not bother to check if it's sending Webmentions to `localhost` or loopback addresses. See the [Security Considerations section][] of the Webmention spec.

Please be aware of these constraints before running lazymention in production. The `domains` option should help with this.

Also, lazymention has no idea how to recurse into subpages of indexes. In the real world, I expect that this will not actually be a real problem.

## Running

lazymention will install a command called, fittingly, `lazymention`. You need to provide, at minimum, a storage location (you can do this on the CLI by passing `-s`), but not configuring anything else will give you a daemon running on port 8000, which is probably no good.

You can provide configuration options in three ways: CLI flags, environment variables, and JSON configuration files - CLI flags override environment variables and environment variables override JSON configs. The default config file location is `/etc/lazymention.json`, but you can override this by passing `-c <path_to_config.json>`. There's a sample file called `lazymention.json.sample` but be sure to actually look at it; lazymention won't work for you if you don't adjust it a little bit.

If you configure with environment variables, just prefix the configuration values with `LAZYMENTION_`. CLI flags are just prefixed with `--`, as you'd expect - just run `lazymention --help` if this is confusing. You can pass more than one value for `--domains` (e.g. `--domains example.com example.net`), as well as passing `--domains` more than once (e.g. `--domains example.com --domains example.net`). If you need to specify more than one domain via environment variables, separate them using `,`.

You should also set `NODE_ENV=production` in the environment, regardless of how you're configuring lazymention.

### Configuration values

Here's what you can configure:

| Name       | Description                                                  | Type          | Default   |
| ---------- | ------------------------------------------------------------ | ------------- | --------- |
| `storage`  | Where to put data. Must be read/writable by the server user. | String        | None      |
| `port`     | Port for the HTTP server to bind to.                         | Number        | 8000      |
| `address`  | Address for the HTTP server to bind to.                      | String        | `0.0.0.0` |
| `logger`   | Whether to have the logger write logs.                       | Boolean       | True      |
| `logfile`  | Where to write logs to.                                      | String        | stdout    |
| `logLevel` | [Bunyan loglevel][] for the logger.                          | String        | `info`  |
| `domains`  | Whitelist of domains to allow in job submissions.            | Array<String> |None      |

The `storage` and `domains` options are required. If you don't want to use a whitelist you can set `domains` to `[]`, but you're _strongly_ discouraged from doing so because of the security implications. In fact, the reason you have to set this manually is to make sure you know what you're doing.

## Using the API

There's exactly one API endpoint, `/jobs/submit`. To trigger lazymention, POST to this endpoint with a JSON payload in the body. The JSON should contain a `url` key whose value is the URL to crawl.

If everything goes well, you'll get back 202 Accepted. If something went wrong, you'll get back 400 Bad Request, unless the domain of the URL you provided isn't in the whitelist in which case you'll get 403 Forbidden.

In the future the reponse will contain a URL that lets you check the status of your job, but in the meantime all you get back is a lame 2xx status code.

### API clients

There's currently one API client, [`ping-lazymention`][]. It's nice for [gulp][] tasks.

Have you written another client? Please send a Pull Request updating this section!

## Embedding

lazymention has an embedding API, in case you want to reuse its functionality in another Node app. You can access both the overall Express application or just the job submission route.

It is safe to concurrently embed unlimited instances of either in the same process so long as they don't share the same data directory.

Note: your app would have to be AGPL 3.0 or later to comply with the licensing terms, but if this is a big problem I might consider relaxing the requirements so just ping me.

### Express application

Available as `require('lazymention').makeApp()`. You need to pass three arguments. In order:

1. A configurations object that contains, at minimum, a `domain` key with the appropriate value
2. A [Bunyan][] `Logger` object (or compatible)
3. A persistence factory function (see below)

Not providing these options as specified will raise `AssertionError`.

The Express app is configured to parse JSON bodies and to compress response bodies. You can't disable the former because it's required for anything useful to happen and you can't disable the latter because that's a bad idea.

`req.log` is configured to be a Bunyan child log and `req.config` is set to the app's config. The API is mounted at `/jobs/submit`, just like the standalone daemon.

### Express `Router`

Available as `require('lazymention').makeRouter()`. You need to pass a single argument, a persistence factory function (see below). Not providing this argument will raise `AssertionError`.

The Router expects certain things:

1. A `req.body` with something useful in it (how to populate this is up to you)
2. A Bunyan `Logger` (or compatible) available as `req.log`
3. A configuration object available as `req.config` and containing, at minimum, a `domains` key

Everything else is up to you.

### Persistence layer

To work with the embedding API you are required to provide some layer for persistence. You can do whatever you want in this layer; the default implementation reads and writes flat files. You can either reuse the default or implement your own.

You can get to the default implementation with `require('lazymention').persistence(<path>)`. Calling this will return a factory function configured with the filesystem `<path>` you provided.

If you don't want to use the default implementation, you need to write a factory function. This function will be passed a single parameter, a child of the  Bunyan `Logger` object you passed. Keep in mind that it will be invoked with a different `Logger` for each request, so connecting out to a database or whatever each and every time this is invoked is probably a terrible idea.

When invoked, your factory function should return an object with two keys, `get` and `set`. `set` must be a function that takes a namespace (string), key (string), an object to set as the value of that key, and a callback with signature `(err<Error>)`. `get` must be a function that takes a namespace, a lookup key and a callback with signature `(err<Error>, data<Object>)`, `data` being the object that was set with `set`.

Values are guaranteed to be JSON-serializable but otherwise there are no restrictions on what they can be.

## Author

AJ Jordan <alex@strugee.net>

## License

AGPL 3.0 or later

 [IndieWeb]: https://indieweb.org/
 [Security Considerations section]: https://www.w3.org/TR/webmention/#security-considerations
 [Bunyan loglevel]: https://github.com/trentm/node-bunyan#levels
 [`ping-lazymention`]: https://github.com/strugee/ping-lazymention/
 [gulp]: https://gulpjs.com/
 [Bunyan]: https://github.com/trentm/node-bunyan


CHANGELOG.md
# lazymention changelog

lazymention follows [Semantic Versioning 2.0.0][semver]. Specifically, the following things are considered to be semver-major if changed in a backwards-incompatible way:

* Configuration value names, semantics, requirements and format
* The public HTTP API
* The embedding API, including the state the router embed expects to have available
* Supported Node.js versions
* This list of items

Changes to everything else are not considered breaking. Notably, these are _not_ considered breaking:

* Log messages

If you think something that isn't in the first list should be covered, or aren't sure about something, please feel free to file an issue and I'll clarify the policy - this list could surely be more precise.

## 2.0.0 - future

### Breaking

* Drop Node 4, 5, and 7 support

## 1.0.1 - 2018-12-06

### Changed

* Various small improvements to documentation

### Fixed

* Fix crash on relative URLs

## 1.0.0 - 2017-12-29

* Initial release

 [semver]: https://semver.org/spec/v2.0.0.html


test/lib/app.js
/*

  Copyright 2017 AJ Jordan <alex@strugee.net>.

  This file is part of lazymention.

  lazymention is free software: you can redistribute it and/or modify it
  under the terms of the GNU Affero General Public License as published
  by the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  lazymention is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public
  License along with lazymention. If not, see
  <https://www.gnu.org/licenses/>.

*/

'use strict';

// TODO test this file

var assert = require('perjury').assert,
    noopLog = require('./log'),
    memoryPersistence = require('./persistence').memoryPersistence,
    _ = require('lodash');

function configureAppSetup(config) {
	return {
		topic: function() {
			var app = require('../../dist/app').makeApp(_.defaults(config, {
				domains: []
			}), noopLog, memoryPersistence()),
			cb = this.callback;

			var server = app.listen(config.port || 5320, config.address || 'localhost', function(err) {
				cb(err, server);
			});
		},
		teardown: function(app) {
			if (app && app.close) {
				app.close(this.callback);
			}
		},
		'it works': function(err, app) {
			assert.ifError(err);
		}
	};
}

function wrapAppSetup(config, obj) {
	if (!obj) {
		obj = config;
		config = {};
	}

	return _.assign({}, configureAppSetup(config), obj);
}

module.exports.wrapAppSetup = wrapAppSetup;


test/lib/log.js
/*

  Copyright 2017 AJ Jordan <alex@strugee.net>.

  This file is part of lazymention.

  lazymention is free software: you can redistribute it and/or modify it
  under the terms of the GNU Affero General Public License as published
  by the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  lazymention is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public
  License along with lazymention. If not, see
  <https://www.gnu.org/licenses/>.

*/

'use strict';

// TODO test this file

var bunyan = require('bunyan');

module.exports = new bunyan({name: 'noop', streams: []});


test/app-test.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

var vows = require('perjury'),
    assert = vows.assert,
    express = require('express'),
    noopLog = require('./lib/log'),
    db = require('../dist/persistence');

vows.describe('app module test').addBatch({
	'When we get the app module': {
		topic: function() {
			return require('../dist/app');
		},
		'it works': function(err) {
			assert.ifError(err);
		},
		'it exports an Express application factory function': function(err, app) {
			assert.isFunction(app.makeApp);
		},
		'it exports an Express Router factory function': function(err, app) {
			assert.isFunction(app.makeRouter);
		},
		'it exports the default persistence layer implementation': function(err, app) {
			assert.isFunction(app.persistence);
		},
		'and we call makeApp()': {
			topic: function(app) {
				return app.makeApp({}, noopLog, db('/tmp'));
			},
			'it works': function(err) {
				assert.ifError(err);
			},
			'it returns an Express application': function(err, app) {
				assert.isFunction(app);
				assert.isFunction(app.listen);
			}
		},
		'and we call makeRouter()': {
			topic: function(app) {
				return app.makeRouter(db(noopLog, '/tmp'));
			},
			'it works': function(err) {
				assert.ifError(err);
			},
			'it exports an Express Router': function(err, router) {
				assert.isFunction(router);
				assert.isTrue(express.Router.isPrototypeOf(router));
			}
		}
	}
}).export(module);


lib/get.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

var _http = require('follow-redirects').http,
    https = require('follow-redirects').https,
    url = require('url'),
    concat = require('concat-stream'),
    pkg = require('../package.json'),
    uuid = require('uuid');

module.exports = function createGetter(_retrievalLog) {
	var retrievalLog = _retrievalLog.child({component: 'get'});

	return function get(path, callback) {
		var opts = url.parse(path),
		    useHTTP = opts.protocol === 'http:',
		    log = retrievalLog.child({job_id: uuid.v4()});

		log.debug({url: path}, 'Retrieving URL with HTTP GET over ' + (useHTTP ? 'HTTP.' : 'TLS.'));

		var http = useHTTP ? _http : https;

		opts.headers = {
			'User-Agent': 'node.js/' + process.versions.node + ' ' + pkg.name + '/' + pkg.version
		};

		var req = http.get(opts);

		req.on('response', function(res) {
			log.debug({res: res}, 'Received response.');
			if (res.statusCode < 200 || res.statusCode >= 300) {
				callback(new Error('Request to ' + path + ' returned HTTP ' + res.statusCode));
				return;
			}

			res.pipe(concat(function(buf) {
				callback(undefined, buf.toString());
			}));
		});

		req.on('error', function(err) {
			log.debug({err: err}, 'Encountered error during request.');
			callback(err);
		});
	};

};


lib/log.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

// XXX test this file

var bunyan = require('bunyan');

module.exports = bunyan.createLogger({
	name: 'lazymention',
	serializers: {
		err: bunyan.stdSerializers.err,
		req: bunyan.stdSerializers.req,
		res: bunyan.stdSerializers.res
	}
});


lib/hentries.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

// XXX if we turn this fully async, make this an EventEmitter or
// something, so you don't have to wait for the array to be fully
// built and can work on individual pages as they're found

var assert = require('assert'),
    url = require('url'),
    mf2 = require('microformat-node'),
    uuid = require('uuid'),
    createGetter = require('./get');

function noop() {}

function _handleNode(baseUrl, thisUrl, node, log, get, enqueue, callback) {
	// XXX should we use more child logs here for better job tracking?
	// TODO error handling is shitty and also basically nonexistant
	assert(node.type.length === 1);

	if (node.type[0] === 'h-feed') {
		log.trace('Found an h-feed.');
		node.children.forEach(function(child) {
			log.trace('Processing h-feed child.');
			var postUrl = child.properties.url;

			assert(postUrl.length <= 1);

			enqueue();

			if (postUrl.length === 1) {
				var resolvedUrl = url.resolve(baseUrl, postUrl[0]);

				log.trace({resolved_url: resolvedUrl}, 'Child URL has length 1; assumed h-entry, resolved to canonical URL and now triggering retrieval.');

				// XXX recognize infinite recursion, where the post links to itself
				get(resolvedUrl, function(err, data) {
					// XXX don't be terrible
					if (err) throw err;

					// XXX move app.js's mf2 parsing into here
					mf2.get({html: data}, function(err, data) {
						if (err) throw err;

						log.trace('Parsed mf2 and processing children.');

						data.items.forEach(function(root) {
							log.trace('Recursively processing child.');
							// We pass noop for `enqueue` because this was already enqueued
							// XXX handle the next URL having not just an h-entry (such that queues get out of sync)
							_handleNode(baseUrl, resolvedUrl, root, log, get, noop, callback);
						});
					});
				});

				return;
			}

			/*
			 We schedule this on the event loop because otherwise we
			 recurse and call callback(node) synchronously. This
			 means that waitingFor gets incremented once and then we
			 fire the callback, which decrements waitingFor back to
			 zero and yields back to the caller.

			 This is obviously unideal since it will always return
			 an array with only the first h-entry encountered. By
			 deferring to the event loop, we give ourselves time to
			 properly call enqueue() for every h-entry encountered,
			 thus preventing waitingFor from being too small.
			*/
			log.trace({length: postUrl.length}, 'Child URL has length >1; triggering recursive processing.');
			process.nextTick(_handleNode.bind(undefined, baseUrl, thisUrl, child, log, get, enqueue, callback));
		});
	} else if (node.type[0] === 'h-entry') {
		log.trace('Found an h-entry; firing callback.');

		assert(node.properties.content.length === 1);

		callback(node, thisUrl);
	} else {
		assert();
	}
}

function makeHentriesHandler(_log) {
	_log = _log.child({component: 'mf2'});

	var get = createGetter(_log);

	return function handleNode(baseUrl, node, callback) {
		var log = _log.child({job_id: uuid.v4()});

		// XXX log `node` too, with a nice Bunyan serializer
		// XXX is this log message misleading? Given that we're really discovering based on a node, not a URL?
		log.info({base_url: baseUrl}, 'Performing h-entry discovery and canonicalization for URL.');

		var arr = [],
		    waitingFor = 0,
		    // I'm suspicious that this won't develop race conditions in the future
		    callbackFired = false;

		_handleNode(baseUrl, baseUrl, node, log, get, function() {
			log.debug('Found and queued an h-entry.');
			waitingFor++;
		}, function(node, resolvedUrl) {
			var result = {
				resolvedUrl: resolvedUrl,
				html: node.properties.content[0].html
			};

			log.debug({resolved_url: resolvedUrl}, 'Processed an h-entry.');

			arr.push(result);

			waitingFor--;
			assert(waitingFor >= 0);

			if (waitingFor === 0) {
				log.debug('Completed all h-entry processing, firing callback.');
				assert(callbackFired === false);
				callbackFired = true;
				callback(undefined, arr);
			}
		});
	};
};

module.exports = makeHentriesHandler;


lib/webmention.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

var cheerio = require('cheerio'),
    webmention = require('send-webmention'),
    _ = require('lodash'),
    uuid = require('uuid'),
    URLToolkit = require('url-toolkit');

module.exports = function createWebmentionDispatcher(_log, db) {
	_log = _log.child({component: 'webmention'});

	return function sendWebmention(sourceUrl, updatedTimestamp, html, cb) {
		var log = _log.child({job_id: uuid.v4()});

		log.info({url: sourceUrl}, 'Processing Webmentions for URL.');

		db.get('PAGE', sourceUrl, function(err, data) {
			if (err) {
				cb(err);
				return;
			}

			if (data.lastSentWebmention > updatedTimestamp) {
				log.debug({
					lastSent: data.lastSentWebmention,
					postTimestamp: updatedTimestamp
				}, 'Ignoring URL because we\'ve seen this post before.');
				cb();
				return;
			}

			var $ = cheerio.load(html);

			var links = $('a'),
			    jobs = 0;

			links.each(function(idx, el) {
				// XXX should we try to check for malformed URLs in a more generic way...?
				if (el.attribs.href.length === 0) {
					log.debug('Dropping attempt to send Webmention to empty href.');
					return;
				}

				// XXX log `el` with a nice Bunyan formatter
				var target = URLToolkit.buildAbsoluteURL(sourceUrl, el.attribs.href);
				log.debug({target: target}, 'Attempting to send Webmention.');
				jobs++;
				webmention(sourceUrl, target, function(err, result) {
					if (err) {
						log.error({target: target, err: err}, 'Error sending Webmention.');
						cb(err);
						return;
					}

					// TODO maybe do something intellingent with result

					log.debug({
						target: el.attribs.href,
						success: result.success,
						status_code: (result.res || {statusCode: undefined}).statusCode
					}, 'Finished Webmention send attempt.');

					jobs--;

					// XXX should we do some sort of db merge instead?
					if (jobs === 0) {
						log.debug('Completed Webmention processing.');
						db.set('PAGE', sourceUrl, _.assign(data, {
							// Normalize to seconds
							lastSentWebmention: Date.now() / 1000
						}), cb);
					}
				});
			});

			if (jobs === 0) {
				log.debug('Nothing to do; completed Webmention processing.');
				process.nextTick(cb);
			}
		});
	};
};


lib/persistence.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

var assert = require('assert'),
    fs = require('fs'),
    path = require('path'),
    slashEscape = require('slash-escape'),
    writeFileAtomic = require('write-file-atomic'),
    uuid = require('uuid');

function computePath(storagePath, key) {
	// 'undefined' here because there was a bug and we have to keep backwards compat
	return path.join(storagePath, slashEscape.escape(key + '_undefined.json'));
}

function createPersistenceDbFactory(storagePath) {
	/*

	  This is a simple filesystem-backed persistence layer for lazymention. Each key
	  corresponds to one file - the filename is set to the slash-escaped key and the
	  value is serialized to JSON and written to the file.

	  The factory is initialized with a filesystem storage path and an in-memory cache
	  backed by a Map. When a write occurs, the cache is updated and a file is written
	  atomically. When a read occurs, the cache is checked and if there's a hit, the
	  callback is scheduled for invocation with the result. If there's a miss, the
	  disk is read. If there's a result the contents of the relevant file are
	  deserialized and passed to the callback; if not an empty object is passed
	  instead. Note that this means that you're always guaranteed to get an object
	  back and it will just be empty if nothing's been set before.

	  Note also that the in-memory cache serves two extremely important functions - it
	  acts as a performance boost, but it also serves as a consistency construct that
	  eliminates the need for any sort of locking. Because Node.js is single-threaded,
	  a write is guaranteed to update the cache in the same tick of the event
	  loop. Any code that would perform a read while the disk write is still in
	  progress will hit the cache and not the disk, and so get back the new, correct
	  result.

	 */

	var cache = new Map();

	return function createPersistenceDb(_dbLog) {
		var dbLog = _dbLog.child({component: 'persistence'});

		return {
			get: function(namespace, _key, cb) {
				assert(storagePath);

				var log = dbLog.child({job_id: uuid.v4()}),
				    key = namespace + '_' + _key;

				log.debug({key: key}, 'Retrieving key.');

				// XXX should we cache empty results?
				if (cache.has(key)) {
					log.trace({key: key}, 'Returning key from in-memory cache.');
					process.nextTick(cb.bind(undefined, undefined, cache.get(key)));
					return;
				}

				fs.readFile(computePath(storagePath, key), function(err, data) {
					// XXX make sure it was the leaf of the directory tree
					if (err && err.code === 'ENOENT') {
						log.debug('Key does not exist; returning empty store.');
						cb(undefined, {});
						return;
					}

					if (err) {
						log.error({err: err}, 'Error retrieving key.');
						cb(err);
						return;
					}

					data = JSON.parse(data.toString());
					log.debug({value: data}, 'Retrieved key.');
					cb(undefined, data);
				});
			},
			set: function(namespace, _key, data, cb) {
				assert(storagePath);

				if (!cb) cb = function() {};

				var log = dbLog.child({job_id: uuid.v4()}),
				    key = namespace + '_' + _key;

				log.debug({key: key, value: data}, 'Setting key.');

				cache.set(key, data);

				writeFileAtomic(computePath(storagePath, key), JSON.stringify(data), function(err) {
					if (err) {
						// XXX should we do something more drastic here? Maybe restore the old value or crash?
						log.error({err: err}, 'Error setting key; in-memory cache is out-of-sync!');
					} else {
						log.debug('Set key.');
					}

					cb(err);
				});
			}
		};
	};
}

module.exports = createPersistenceDbFactory;


lib/app.js
/*

Copyright 2017 AJ Jordan <alex@strugee.net>.

This file is part of lazymention.

lazymention is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

lazymention is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with lazymention. If not, see
<https://www.gnu.org/licenses/>.

*/

'use strict';

var express = require('express'),
    compression = require('compression'),
    _ = require('lodash'),
    uuid = require('uuid'),
    crawl = require('crawl-mf2'),
    url = require('url'),
    assert = require('assert'),
    createWebmentionDispatcher = require('./webmention'),
    createGetter = require('./get'),
    persistence = require('./persistence'); // Not used, only exported

function makeRouter(createDb) {
	assert(typeof createDb === 'function');

	var router = express.Router();

	router.use(function(req, res, next) {
		req.db = createDb(req.log);
		next();
	});

	router.post('/submit', function(req, res, next) {
		var webmention = createWebmentionDispatcher(req.log, req.db),
		    get = createGetter(req.log);

		req.log.info({req: req}, 'Handling incoming request.');

		// TODO assign a URL for status checks

		if (_.isEmpty(req.body)) {
			res.status(400);
			res.end('JSON object was empty, or wrong Content-Type provided');
			req.log.info({res: res}, 'Returning error due to malformed request JSON.');
			return;
		}

		var target = req.body.url;

		if (!target) {
			res.status(400);
			res.end('No URL provided');
			req.log.info({res: res}, 'Returning error due to missing URL in request.');
			return;
		}

		var hostname = url.parse(target).hostname;
		if (req.config.domains.length > 0 && !req.config.domains.includes(hostname)) {
			res.status(403);
			res.end('Domain ' + hostname + ' not in server whitelist');
			req.log.info({res: res, domain: hostname}, 'Returning Forbidden due to domain not in whitelist.');
			return;
		}

		res.status(202);

		req.log.info({url: target}, 'Triggered Webmention processing from inbound request.');

		// XXX UA
		var crawler = crawl(target);

		crawler.on('error', function(err) {
			req.warn({err: err}, 'Error while crawling microformats2.');
		});

		// XXX update status throughout this bit
		crawler.on('h-entry', function(url, node) {
			req.log.info('Found canonical h-entry; triggering Webmention processing.');
			// XXX pass an updated timestamp
			webmention(url, 0, node.properties.content[0].html, function(err) {
				if (err) {
					req.log.error({err: err}, 'Error during Webmention processing.');
					// XXX update status
					return;
				}

				// TODO update status check URL, etc.
				req.log.info('Finished Webmention processing.');
			});
		});

		res.end();
		req.log.info({res: res}, 'Jobs queued; returning success.');
	});

	return router;
}

function makeApp(config, _log, createDb) {
	assert(typeof config === 'object');
	assert(typeof _log === 'object');
	assert(typeof _log.info === 'function');
	assert(typeof createDb === 'function');

	var app = express(),
	    router = makeRouter(createDb);

	app.use(compression());
	app.use(express.json());

	app.use(function(req, res, next) {
		// XXX pass this into modules
		req.log = _log.child({req_id: uuid.v4()});
		req.config = config;
		next();
	});

	app.use('/jobs/', router);

	return app;
}

module.exports.makeApp = makeApp;
module.exports.makeRouter = makeRouter;
module.exports.persistence = persistence;
